### ISSUE TITLE
Unit testing functions that use workerize'd imports

### ISSUE THREAD

**Issue by wbobeirne** [developit/workerize-loader#48]
Has anyone come up with a way to write unit tests (e.g. jest, mocha) for functions that user workerize-loader imports? It'd be great if there was an easy way to replace the default workerize-loader behavior with a simple promise wrapper when in a testing environment. Any advice would be appreciated.

**Comment by developit** [comment]
If you export async functions from the module you're workerizing, they are callable without workerize involved:

```js
export async function foo() {
  return 42;
}
```

```js
import { foo } from 'workerize-loader!./a'
foo()  // Promise<42>
```

```js
import { foo } from './a'
foo()  // Promise<42>
```

**Comment by wbobeirne** [comment]
Yeah, that works like a charm for testing the worker functions themselves. The issue I have is that the function I want to unit test is in a file that _includes that worker_, or a file that includes that file etc. etc.:

```
import { foo } from 'workerize-loader!./a'

function async iWantToUnitTestThis() {
  const number = await foo();
  return foo * 2;
}
```

I hacked around this by using [mock-require](https://www.npmjs.com/package/mock-require) and mocking the worker for each of those files before running my tests:

```
function makeWorkerFactory(module) {
  return () => module
}

const workers = ['list', 'of', 'worker.js', 'files']
workers.forEach(worker => {
  const source = require(worker)
  return mock(path, makeWorkerFactory(source))
})
```

But obviously this isn't ideal.

**Comment by developit** [comment]
Is the issue that your running your test directly in Node so there is no Worker support?

If so, just drop jsdom-worker in there:
https://github.com/developit/jsdom-worker

**Comment by danreeves** [comment]
The issue is that workerize-loader changes the shape of the module. You can't use a module that imports a workerized module in jest, for example, because the default export doesn't exist unless you've run it through webpack first.

**Comment by serdaroquai** [comment]
an alternative way is to tell jest to map the module to a mock implementation
```javascript
"moduleNameMapper": {
  "workerize-loader!../../worker/some.worker":"<rootDir>/src/worker/some.worker.mock.js"
}
```
to `jest.config.js` or if you are using create-react-app to `package.json`.   

...and supply a mock worker at the path you described
```javascript
module.exports = () => require('./some.worker');
```
Reference: [https://twitter.com/james_k_nelson/status/1175259347422564357](https://twitter.com/james_k_nelson/status/1175259347422564357)

**Comment by developit** [comment]
I added a variant of that solution to a [new testing section in the readme](https://github.com/developit/workerize-loader/blob/master/README.md#testing). Cheers!

**Comment by paulmelero** [comment]
The `transform` wasn't working for me either (as in #103). This is the error I was getting:

> `Cannot find module 'workerize-loader!~/core/referenceStyler'`

![image](https://user-images.githubusercontent.com/6775220/90435353-29102f00-e0cf-11ea-84d4-c5b3d34d2691.png)

I opted for using the `moduleNameMapper` as @serdaroquai [suggested](https://github.com/developit/workerize-loader/issues/48#issuecomment-557414099).

Although I noticed it might be a little bit slow, the test is taking now 5s. Not sure where the problem is yet.

Anyway, thank you for this great library! 🙌 ❤️
